Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+
+            // Fill the track state
+            trackStateProxy.predicted() = boundParams.parameters();
+            trackStateProxy.predictedCovariance() = *boundParams.covariance();
+            trackStateProxy.jacobian() = jacobian;
+            trackStateProxy.pathLength() = pathLength;
+          } else {
+            ACTS_VERBOSE("Detected in-sensitive surface " << surface->geoID());
+
+            // Transport & get curvilinear state instead of bound state
+            auto [curvilinearParams, jacobian, pathLength] =
+                stepper.curvilinearState(state.stepping, true);
+
+            // Set the surface
+            trackStateProxy.setReferenceSurface(
+                Surface::makeShared<PlaneSurface>(
+                    curvilinearParams.position(),
+                    curvilinearParams.momentum()));
+
+            // Fill the track state
+            trackStateProxy.predicted() = curvilinearParams.parameters();
+            trackStateProxy.predictedCovariance() =
+                *curvilinearParams.covariance();
+            trackStateProxy.jacobian() = jacobian;
+            trackStateProxy.pathLength() = pathLength;
+          }
+
+          // Set the filtered parameter index to be the same with predicted
+          // parameter
+          trackStateProxy.data().ifiltered = trackStateProxy.data().ipredicted;
+
+          // Check if need to stop this branch
+          if (not m_branchStopper(result.fittedStates, result.currentTip)) {
+            // Remember the active tip and its state
+            result.activeTips.emplace(result.currentTip, std::move(tipState));
+            // Count the valid branches on current surface
+            nBranchesOnSurface++;
+          }
+        } else {
+          // No state is created, this branch is still valid. Count the
+          // branch on current surface
+          nBranchesOnSurface++;
+        }
+        // Update state and stepper with material effects
+        materialInteractor(surface, state, stepper, fullUpdate);
+      } else {
+        // Neither measurement nor material on surface, this branch is still
+        // valid. Count the branch on current surface
+        nBranchesOnSurface++;
+      }
+
+      // Reset current tip if there is no branch on current surface
+      if (nBranchesOnSurface == 0) {
+        ACTS_DEBUG("Branch on surface " << surface->geoID() << " is stopped");
+        if (not result.activeTips.empty()) {
+          result.currentTip = result.activeTips.rbegin()->first;
+          ACTS_VERBOSE(
+              "Propagation jumps to branch with tip = " << result.currentTip);
+          reset(state, stepper, result);
+        } else {
+          ACTS_VERBOSE("Stop forward Kalman filtering with "
+                       << result.trackTips.size() << " found tracks");
+          result.forwardFiltered = true;
+        }
+      }
+
+      return Result<void>::success();
+    }
```

Comment:
This method is **massive** (which makes sense because it does a lot of sophisticated stuff). Is there any way you could maybe break this down intro smaller methods?

Reply:
Right. I had exactly this plan! The creation of track state has now been handled by separate methods (but could still look massive...)